algosdk/abi.py

```diff
@@ -0,0 +1,570 @@
+            return f"string"
+        elif self.abi_type_id == BaseType.Tuple:
+            if not self.child_types:
+                return f"()"
+            return f"({','.join(str(t) for t in self.child_types)})"
+        else:
+            raise error.ABITypeError("failed to infer ABI type from id")
+
+    def __eq__(self, other) -> bool:
+        if (
+            not isinstance(other, Type)
+            or self.abi_type_id != other.abi_type_id
+            or self.precision != other.precision
+            or self.bit_size != other.bit_size
+            or self.static_length != other.static_length
+            or len(self.child_types) != len(other.child_types)
+        ):
+            return False
+        for i in range(len(self.child_types)):
+            if self.child_types[i] != other.child_types[i]:
+                return False
+        return True
+
+    @staticmethod
+    def make_uint_type(type_size):
+        if (
+            not isinstance(type_size, int)
+            or type_size % 8 != 0
+            or type_size < 8
+            or type_size > 512
+        ):
+            raise error.ABITypeError(f"unsupported uint bitSize: {type_size}")
+        return Type(BaseType.Uint, bit_size=type_size)
+
+    @staticmethod
+    def make_byte_type():
+        return Type(BaseType.Byte)
+
+    @staticmethod
+    def make_ufixed_type(type_size, type_precision):
+        if (
+            not isinstance(type_size, int)
+            or type_size % 8 != 0
+            or type_size < 8
+            or type_size > 512
+        ):
+            raise error.ABITypeError(
+                f"unsupported ufixed bitSize: {type_size}"
+            )
+        if (
+            not isinstance(type_precision, int)
+            or type_precision > 160
+            or type_precision < 1
+        ):
+            raise error.ABITypeError(
+                f"unsupported ufixed precision: {type_precision}"
+            )
+        return Type(
+            BaseType.Ufixed, bit_size=type_size, precision=type_precision
+        )
+
+    @staticmethod
+    def make_bool_type():
+        return Type(BaseType.Bool)
+
+    @staticmethod
+    def make_static_array_type(arg_type, array_len):
+        assert isinstance(arg_type, Type)
+        return Type(
+            BaseType.ArrayStatic,
+            child_type=arg_type,
+            child_types=list(),
+            static_length=array_len,
+        )
+
+    @staticmethod
+    def make_address_type():
+        return Type(BaseType.Address)
+
+    @staticmethod
+    def make_dynamic_array_type(arg_type):
+        assert isinstance(arg_type, Type)
+        return Type(
+            BaseType.ArrayDynamic, child_type=arg_type, child_types=list()
+        )
+
+    @staticmethod
+    def make_string_type():
+        return Type(BaseType.String)
+
+    @staticmethod
+    def make_tuple_type(arg_types):
+        if len(arg_types) >= 2 ** 16:
+            raise error.ABITypeError(f"tuple args exceed 2^16")
```

Comment:
more specifically, should say something like tuple element number larger than uint16 max.